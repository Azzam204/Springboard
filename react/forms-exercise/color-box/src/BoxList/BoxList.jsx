import { React, useState } from "react";
import Box from "../Box/Box";
import NewBoxForm from "../NewBoxForm/NewBoxForm";
import { v4 as uuid } from "uuid"
import "../App.css"

function BoxList() {
  const [boxes, setBoxes] = useState([])

  const addBox = box => {
    let newBox = { ...box, id: uuid() };
    setBoxes(boxes => [...boxes, newBox]);
  }

  const delBox = box => {
    setBoxes(boxes.filter(b => b !== box));
  }


  return (
    <>
      <div className="BoxList">
        {boxes.map(box => (
          <Box key={box.id}
            color={box.color}
            height={box.height}
            width={box.width}
            delBox={() => delBox(box)}
          />
        ))}
      </div>
      <NewBoxForm addBox={addBox} />
    </>
  )
}

export default BoxList