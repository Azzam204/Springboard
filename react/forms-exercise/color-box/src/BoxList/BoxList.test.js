import { fireEvent, render } from "@testing-library/react"
import '@testing-library/jest-dom'
import BoxList from "./BoxList"

// Smoke test

it('renders without crashing', function () {
  render(
    <BoxList />
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <BoxList />
  );
  expect(asFragment()).toMatchSnapshot();
});

it("can add a new box", function () {
  const { queryByPlaceholderText, queryByText } = render(<BoxList />);

  expect(queryByText('x')).not.toBeInTheDocument();

  const colorInput = queryByPlaceholderText("Box color");
  const heightInput = queryByPlaceholderText("Box height");
  const widthInput = queryByPlaceholderText("Box width");
  const submitBtn = queryByText("Add Box!")

  fireEvent.change(colorInput, { target: { value: 'red' } });
  fireEvent.change(heightInput, { target: { value: '100' } });
  fireEvent.change(widthInput, { target: { value: '100' } });
  fireEvent.click(submitBtn);

  expect(queryByText('x')).toBeInTheDocument();

})


it("can remove a box", function () {
  const { queryByPlaceholderText, queryByText } = render(<BoxList />);

  expect(queryByText('x')).not.toBeInTheDocument();

  const colorInput = queryByPlaceholderText("Box color");
  const heightInput = queryByPlaceholderText("Box height");
  const widthInput = queryByPlaceholderText("Box width");
  const submitBtn = queryByText("Add Box!");

  fireEvent.change(colorInput, { targer: { value: 'red' } });
  fireEvent.change(heightInput, { target: { value: '100' } });
  fireEvent.change(widthInput, { target: { value: '100' } });
  fireEvent.click(submitBtn);

  expect(queryByText('x')).toBeInTheDocument();

  const delBtn = queryByText('x');

  fireEvent.click(delBtn);

  expect(queryByText('x')).not.toBeInTheDocument();

})