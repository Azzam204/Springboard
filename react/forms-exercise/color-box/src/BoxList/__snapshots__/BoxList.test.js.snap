// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  />
  <form
    class="NewBoxForm"
  >
    <div>
      <input
        id="color"
        name="color"
        placeholder="Box color"
        type="color"
        value="red"
      />
    </div>
    <div>
      <input
        id="height"
        name="height"
        placeholder="Box height"
        type="number"
        value=""
      />
    </div>
    <div>
      <input
        id="width"
        name="width"
        placeholder="Box width"
        type="number"
        value=""
      />
    </div>
    <button>
      Add Box!
    </button>
  </form>
</DocumentFragment>
`;
